### PR TITLE
Prevent pythran from having a permanent side effect on CC and CXX

### PR DESCRIPTION
**Pull request by serge-sans-paille**
Solved by winning some king of monkey-patching race with numpy, cython
and setuptools...

### CODE REVIEW

**Comment by ashwinvis** [comment]
I don't know what to say! If we invert the order of imports will it work (see below)?

```python
from pythran.dist import PythranBuildExt
from Cython.Distutils.build_ext import build_ext
```
I hope we are not trading one side effect for another :smile:

**Comment by serge-sans-paille** [comment]
> I don't know what to say! If we invert the order of imports will it work (see
> below)?

build_ext is monkey-patched by pythran, so the user should not do that
kind of things :-)

**Comment by serge-sans-paille** [comment]
but yeah, i'm not happy with that either...

File: pythran/dist.py [hunk 1]
```diff
@@ -4,13 +4,48 @@
...
+
+                try:
+                    return LegacyBuildExt.build_extension(self, ext)
+                finally:
+                    for attr in prev.keys():
+                        getattr(self.compiler, attr)[0] = prev[attr]
+            else:
+                return LegacyBuildExt.build_extension(self, ext)
+
+    build_ext.build_ext = PythranBuildExt
```
Comment (ashwinvis) [inline pythran/dist.py]: Instead of monkey-patching, leave it to the user to inherit the custom build_ext class?

File: pythran/dist.py [hunk 2]
```diff
@@ -4,13 +4,48 @@
...
+                    prev[attr] = getattr(self.compiler, attr)[0]
+                    getattr(self.compiler, attr)[0] = ext.cxx
+
+                try:
+                    return LegacyBuildExt.build_extension(self, ext)
+                finally:
+                    for attr in prev.keys():
+                        getattr(self.compiler, attr)[0] = prev[attr]
+            else:
+                return LegacyBuildExt.build_extension(self, ext)
```
Comment (ashwinvis) [inline pythran/dist.py]: Maybe use `super(PythranBuildExt, self).build_extension(self, ext)`  here and above?

File: pythran/config.py [hunk 3]
```diff
@@ -99,8 +105,9 @@ def parse_define(define):
         sys.stdout = old_stdout
 
     # final macro normalization
-    extension["define_macros"] = [parse_define(dm) for dm in
-                                  extension["define_macros"]]
+    extension["define_macros"] = [
+        dm if isinstance(dm, tuple) else parse_define(dm)
+        for dm in extension["define_macros"]]
     return extension
```
Comment (ashwinvis) [inline pythran/config.py]: Function `compiler()` a few lines below still relies on `os.environ.get('CXX'...`

File: pythran/dist.py [hunk 4]
```diff
@@ -20,6 +55,9 @@ class PythranExtension(Extension):
     They must be processable by pythran, of course.
 
     The compilation process ends up in a native Python module.
+
+    Unlike distutils.core.Extension it also accepts a cxx keyword argument to
+    force the compiler executable.
```
Comment (ashwinvis) [inline pythran/dist.py]: This is a good feature. If only it were implemented upstream, we could have avoided all these hacks :)

File: pythran/config.py [hunk 5]
```diff
@@ -72,13 +70,21 @@ def parse_define(define):
...
     # using / as separator as advised in the distutils doc
     extension["include_dirs"].append(here + '/pythran')
+
+    extra.pop('language', None)  # forced to c++ anyway
+    cxx = extra.pop('cxx', None)
+    if cxx is not None:
+        extension['cxx'] = cxx
+    elif cfg.get('compiler', 'CXX'):
+        extension['cxx'] = str(cfg.get('compiler', 'CXX'))
+
```
Comment (ashwinvis) [inline pythran/config.py]: It maybe useful to add some sort of warning to the user (`DeprecationWarning` ?), to use the custom build_ext class when `CXX` is configured.

**Comment by ashwinvis** [comment]
Updated this branch: https://github.com/fluiddyn/pythran/commit/bcc45e05317b6d9775a4bc207461f4f8705c8093

**Comment by serge-sans-paille-qb** [comment]
Obsoleted by #940

**Pull request closed without merging**